src/mlpack/core/data/load_impl.hpp

```diff
@@ -20,376 +21,303 @@
+    {
+      Log::Info << "Size is " << matrix.n_rows << " x "
+          << matrix.n_cols << ".\n";
+    }
   }
 
   Timer::Stop("loading_data");
 
-  // Finally, return the success indicator.
   return success;
 }
 
-// Load with mappings.  Unfortunately we have to implement this ourselves.
-template<typename eT, typename PolicyType>
-bool Load(const std::string& filename,
-          arma::Mat<eT>& matrix,
-          DatasetMapper<PolicyType>& info,
-          const bool fatal,
-          const bool transpose)
+template<typename MatType>
+bool LoadDense(const std::string& filename,
+               MatType& matrix,
+               TextOptions& opts,
+               std::fstream& stream)
 {
-  // Get the extension and load as necessary.
-  Timer::Start("loading_data");
-
-  // Get the extension.
-  std::string extension = Extension(filename);
-
-  // Catch nonexistent files by opening the stream ourselves.
-  std::fstream stream;
-  stream.open(filename.c_str(), std::fstream::in);
-
-  if (!stream.is_open())
-  {
-    Timer::Stop("loading_data");
-    if (fatal)
-      Log::Fatal << "Cannot open file '" << filename << "'. " << std::endl;
-    else
-      Log::Warn << "Cannot open file '" << filename << "'; load failed."
-          << std::endl;
-
-    return false;
-  }
+  bool success;
+  if (opts.Format() != FileType::RawBinary)
+    Log::Info << "Loading '" << filename << "' as " 
+        << opts.FileTypeToString() << ".  " << std::flush;
 
-  if (extension == "csv" || extension == "tsv" || extension == "txt")
+  // We can't use the stream if the type is HDF5.
+  if (opts.Format() == FileType::HDF5Binary)
   {
-    Log::Info << "Loading '" << filename << "' as CSV dataset.  " << std::flush;
-    try
-    {
-      LoadCSV loader(filename);
-      loader.LoadCategoricalCSV(matrix, info, transpose);
-    }
-    catch (std::exception& e)
-    {
-      Timer::Stop("loading_data");
-      if (fatal)
-        Log::Fatal << e.what() << std::endl;
-      else
-        Log::Warn << e.what() << std::endl;
-
-      return false;
-    }
+    success = LoadHDF5(filename, matrix, opts);
   }
-  else if (extension == "arff")
+  else if (opts.Format() == FileType::CSVASCII)
   {
-    Log::Info << "Loading '" << filename << "' as ARFF dataset.  "
-        << std::flush;
-    try
-    {
-      LoadARFF(filename, matrix, info);
+    success = LoadCSVASCII(filename, matrix, opts);
 
-      // We transpose by default.  So, un-transpose if necessary...
-      if (!transpose)
-      {
-        return inplace_transpose(matrix, fatal);
-      }
-    }
-    catch (std::exception& e)
-    {
-      Timer::Stop("loading_data");
-      if (fatal)
-        Log::Fatal << e.what() << std::endl;
-      else
-        Log::Warn << e.what() << std::endl;
-
-      return false;
-    }
+    if (matrix.col(0).is_zero())
+      Log::Warn << "First column is zeros, if dataset contains headers, "
+                   "please use HasHeaders() option." << std::endl;
```

Comment:
```suggestion
      Log::Warn << "data::Load(): the first line in '" << filename << "' was "
          << "loaded as all zeros; if the first row is headers, specify "
          << "`HasHeaders() = true` in the given DataOptions." << std::endl;
```

I tried to modify the text so that a user will know where the message came from, what file was being referenced, and what to do about it; let me know what you think.

Also, since we don't yet document `HasHeaders()`, do you think we should leave this check commented out for now?